File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int dx[] = {1,-1,0,0};
int dy[] = {0,0,1,-1};
int visit[6][6];
struct xy{
    int x;
    int y;
    int cnt;
};

bool bfs(vector<string> map,int i,int j){
    queue<xy> q;
    q.push({i,j, 0});
    visit[i][j] = 1;
    
    while(q.size()!=0){
        int x = q.front().x;
        int y = q.front().y;
        int cnt = q.front().cnt;
        if(cnt > 1)
            break;
        
        q.pop();
        for(int k=0;k<4;k++){
            int nx = x + dx[k];
            int ny = y + dy[k];
            if(nx < 0 or nx >= map.size() or ny < 0 or ny >= map[0].size() or visit[nx][ny])
                continue;
            visit[nx][ny] = 1;
            if(map[nx][ny] == 'P')
                return false;
            if(map[nx][ny] == 'O')
                q.push({nx,ny, cnt + 1});
        }
    }
    
    return true;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for(int i=0;i<places.size();i++){
        bool suc = true;
        
        for(int j=0;j<5;j++)
            for(int k=0;k<5;k++)
                visit[j][k] = 0;
        
        for(int j=0;j<places[i].size();j++){
            for(int k=0;k<places[i][j].size();k++){
                if(places[i][j][k] == 'P')
                    suc = bfs(places[i],j,k);
                if(!suc)
                    break;
            }
            if(!suc)
                break;
        }
        if(!suc)
            answer.push_back(0);
        else
            answer.push_back(1);
    }
    return answer;
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.io.*;
import java.util.*;
class Solution {
    static char map[][][];
    static int visit[][];
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        
        visit = new int[5][5];
        map = new char[5][5][5];
        for(int i = 0; i < 5; i++)
            for(int j = 0; j < 5; j++)
                for(int k = 0; k < 5; k++) map[i][j][k] = places[i][j].charAt(k);
        
        for(int i = 0; i < 5; i++){
            int flag = 1;
            for(int j = 0; j < 5; j++){
                for(int k = 0; k < 5; k++){
                    if(map[i][j][k] != 'P') continue;
                    if(!bfs(i,j,k)){
                        flag = 0; break;
                    }
                }if(flag == 0) break;
            }
            answer[i] = flag;
        }
        
        return answer;
    }
    static boolean bfs(int i, int j, int k){
        Queue<xy> q = new LinkedList<>();
        q.add(new xy(j,k,0));
        
        for(int l = 0; l < 5; l++) Arrays.fill(visit[l], 0);
        visit[j][k] = 1;
        
        while(!q.isEmpty()){
            xy x = q.poll();
            if(x.cnt == 2) continue;
            
            for(int l = 0; l < 4; l++){
                int nx = x.x + dx[l];
                int ny = x.y + dy[l];
                if(nx < 0 || nx >= 5 || ny < 0 || ny >= 5) continue;
                if(visit[nx][ny] == 1) continue;
                if(map[i][nx][ny] == 'X') continue;
                if(map[i][nx][ny] == 'P') return false;
                visit[nx][ny] = 1;
                q.add(new xy(nx, ny, x.cnt + 1));
            }
        }
        
        return true;
    }
    
    static class xy{
        int x, y, cnt;
        xy(int a, int b,int c){
            x = a; y = b; cnt = c;
        }
    }
}